jax/interpreters/pxla.py

```diff
@@ -345,30 +332,56 @@ def block_until_ready(self):
   @property
   def _value(self):
     if self._npy_value is None:
-      ids = self._ids()
+      # TODO(skye): remove this to avoid transferring replicated buffers?
       self.copy_to_host_async()
-      self._npy_value = self._collect([self.device_buffers[i].to_py() for i in ids])
+      self._npy_value = onp.empty(self.aval.shape, self.aval.dtype)
+      # TODO(skye): benchmark and possibly switch to a set (maybe with with
+      # cached hashable indices?)
+      already_copied_indices = []
+      for buf, idx in zip(self.device_buffers, self.logical_indices):
+        if idx in already_copied_indices:
+          continue
+        self._npy_value[idx] = buf.to_py()
+        already_copied_indices.append(idx)
     return self._npy_value
```

Comment:
Related to the "validity constraints" comment, I think this method is relying on the `self.logical_indices` forming a cover of the logical array (otherwise we might return uninitialized memory to the user here).